fix(ngValue): set value property instead of attr

m-amr opened the pull request:
in case of input element set value using property instead of attribute
fixed #13984

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1756,12 +1756,25 @@ var ngValueDirective = function() {
+           *  so we set both the value attribute and property here to avoid this problem.
+           */
+          attr.$set('value', value);
+          elm[0].value = value;
```
Could you extract these two lines (along with the comment above) into a helper function to avoid duplication.

gkalpak commented:
The docs also need to be updated to reflect the changes.

I would still add a breaking change note, since we are breaking a (documented-ish) assumption.
I can't think of any usecase, where this change would indeed break an app, but it is a breaking change.

m-amr commented:
@gkalpak 

1. I have extracted a method to update the value to prevent code duplication .
2. I have updating the commit message with the breaking change.
3. I can't found anything to be update in the documentation. 
because the default behaviour of the ng-value is to update the input when the model change and this is a bug. please tell me if documentation is need to be updated ? 

Thanks for your feedback.

gkalpak commented:
WRT to the docs, e.g. soewhere it says that the resulting _"value will be bound to the `value` attribute of the input element_", which is not accurate anymore.

The docs need a more thorough make-over anyway, because they mention `option` and `input[radio]` elements, but with this change `ngValue` can be used for more purposes, so I can take care of the docs update while mergin or later.

--
It LGTM.
@Narretz, since you were working on some `ngValue` stuff, do you have any objection/concerns about this ?

Narretz commented:
I would like to see the setter code to be more precise in terms of which elements need attribute set and which need property set. So basically the setter fn should be set up based on the element type. I assume the value property should be set for textarea and input[text], while option and input[radio|checkbox] are good with attribute only.
And the docs must be updated. Atm, the docs are pretty explicit that the directive is only for radio/checkbox and option elements.
So the change is actually a feature and not a fix.

m-amr commented:
@Narretz I think that the setter is setting the property in case of boolean values, i tried ti fix it in this pull request  https://github.com/angular/angular.js/pull/14023

Narretz commented:
@m-amr I'm not sure I follow. Input elements / controls cannot have boolean values. Their values are always string values. Angular models are a different situation, but they generally follow this rule, with exceptions.

m-amr commented:
@Narretz 
sorry I didn't mean that their values are boolean, i mean that they had properties that is boolean values

A number of attributes are boolean attributes. The presence of a boolean attribute on an element represents the true value, and the absence of the attribute represents the false value.

        var BOOLEAN_ATTR = {};
           forEach('multiple,selected,checked,disabled,readOnly,required,open'.split(','), function(value) {
             BOOLEAN_ATTR[lowercase(value)] = value;
         });
        var BOOLEAN_ELEMENTS = {};
        forEach('input,select,option,textarea,button,form,details'.split(','), function(value) {
            BOOLEAN_ELEMENTS[value] = true;
        });

so if the in case of boolean attribute it sets the property

         $set: function(key, value, writeAttr, attrName) {
        // TODO: decide whether or not to throw an error if "class"
        //is set through this function since it may cause $updateClass to
        //become unstable.

        var node = this.$$element[0],
            booleanKey = getBooleanAttrName(node, key),
            aliasedKey = getAliasedAttrName(key),
            observer = key,
            nodeName;

        if (booleanKey) {
          this.$$element.prop(key, value);
          attrName = booleanKey;
        } else if (aliasedKey) {
          this[aliasedKey] = value;
          observer = aliasedKey;
        }

so it will call this.$$element.prop(key, value); in case of boolean attribute.

m-amr commented:
Sorry i had done git --amend to wrong commit, i have reset it and push it.

m-amr committed:
fix(ngValue): set value property and attribute instead of setting attribute only
input element reads the value from value property and override the value attribute so we need to set both

Closes #14031
Closes #13984

BREAKING CHANGE:

NgValue is no longer updating the "value attribute" only when the model change,
because input element reads the value from "value attribute" only in case of "value property" has no value
but when "value property" changed the input element will stop reading the value from "value attribute" and
read it form the "value property".
so the input value in the gui will not equal the value of the model, becuase the model update the "value attribute" only.

Now ngValue updates both of "value attribute" and "value property" when the model change
This shouldn't affect any application becuase this is a bug and input value in the gui should be updated when the model change

Narretz commented on src/ng/directive/input.js:
```diff
@@ -1750,24 +1750,33 @@ var CONSTANT_VALUE_REGEXP = /^(true|false|\d+)$/;
+     *  so we set both the value attribute and property here to avoid this problem.
+     */
+    attr.$set('value', value);
+    element[0].value = value;
```
Should use `prop()`

Narretz commented on test/ng/directive/inputSpec.js:
```diff
@@ -3046,6 +3046,26 @@ describe('input', function() {
+      expect(element[0].value).toBe('something');
+      expect(element[0].getAttribute('value')).toBe('something');
+
+      helper.changeInputValueTo('newValue');
```
There is no expect after his change, so either it's redundant or an expect is missing

m-amr replied:
This is not 'redundant' it important to test that the 
"input binding to the model" is still working after user edited the input, without it we are not testing the Issue 
if "value property"  has no value, it will read the value from "value attribute" 
The Issue happened when user type something in the input,
see the following step
1. User edit the input
2 . then change the model value programmatically by changing the scope.
3. The input is not updated will the new value of the model

I am trying to simulating what happen in the Issue https://github.com/angular/angular.js/issues/13984

do i need to add expect after it ?

Narretz replied:
Yes, you should add an expect. Otherwise you can't be sure that the input / scope has really been changed.

gkalpak replied:
@m-amr, basically this is redundant in this test, because we have already tested that the `value` property _does_ have a value. If you wanted to test the scenario your describe, you should edit the input, before setting the value (maybe in a separate test).

<sub>FWIW, I don't think an extra `expect` is needed.</sub>

m-amr commented:
@Narretz @gkalpak 
Could you please give it another look, i have done the following
 1. use prop function to set the value instead of element.value
 2. remove redundant code from test
 3. i had updated the docs

Thanks.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1701,6 +1701,11 @@ var CONSTANT_VALUE_REGEXP = /^(true|false|\d+)$/;
  * the bound value.
  *
+ * It can be used to achieve one-way binding of a given expression to the value of
+ * {@link input[text] `input[text]`} or {@link textarea `textarea`}. so that when the expression change
```
Why only `input[text]`. Won't it work with any input type ?

m-amr replied:
I have to change it to be input type

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1711,7 +1716,7 @@ var CONSTANT_VALUE_REGEXP = /^(true|false|\d+)$/;
  * @element input
  * @param {string=} ngValue angular expression, whose value will be bound to the `value` attribute
- *   of the `input` element
+ * and `value` property of the `input` or `textarea` element
```
I don't think `input` was meant as `<input>` (it would be wrong). Maybe we should just say `element`, since it can be `input`, `textarea`, `option` (anything actually).

m-amr replied:
I have change it, i am now using 'element'  instead of input and textarea

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -3046,6 +3046,22 @@ describe('input', function() {
 
+    they('should update the $prop "value" property and attribute after change the "value" property', {
+      input: '<input type="text" ng-value="value">',
+      textarea: '<textarea ng-value="value">></textarea>'
```
There's an extra `>`

m-amr replied:
i had fixed it fixed.

gkalpak commented:
> I would like to see the setter code to be more precise in terms of which elements need attribute set and which need property set. So basically the setter fn should be set up based on the element type. I assume the value property should be set for textarea and input[text], while option and input[radio|checkbox] are good with attribute only.

I'd rather set both and not worry about browser quirks (unless this "strategy" would lead to problems).
@Narretz, do you know of any usecase, where setting both would be a problem ?

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1701,6 +1701,10 @@ var CONSTANT_VALUE_REGEXP = /^(true|false|\d+)$/;
  * the bound value.
  *
+ * It can be used to achieve one-way binding of a given expression to the value of
+ * {@link input `input`}. so that when the expression change the element is set to the bound value.
```
`.`  -->  `,`

m-amr replied:
Done.

Narretz commented:
@gkalpak I don't know of any way this could cause problems. I was just conservative, I guess. 
The BC notice needs some love, because it's a bit hard to understand, otherwise LGTM.
